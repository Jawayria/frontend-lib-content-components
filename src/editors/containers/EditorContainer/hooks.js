import { useSelector } from 'react-redux';

import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import * as appHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const handleSaveClicked = ({ getContent, dispatch }) => (
  () => saveBlock({ content: getContent(), dispatch })
);
export const handleCancelClicked = ({ onClose }) => {
  if (onClose) {
    return onClose;
  }
  return navigateCallback(useSelector(selectors.app.returnUrl));
};
export const isInitialized = () => useSelector(selectors.app.isInitialized);
export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));
