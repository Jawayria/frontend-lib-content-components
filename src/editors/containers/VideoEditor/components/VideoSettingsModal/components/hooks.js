import { useSelector } from 'react-redux';

import { keyStore } from '../../../../../utils';
import { actions, selectors } from '../../../../../data/redux';

export const selectorKeys = keyStore(selectors.video);

export const widgetValue = (key, dispatch) => ({
  formValue: useSelector(selectors.video[key]),
  setFormValue: (val) => dispatch(actions.video.load({ [key]: val })),
});

export default {
  selectorKeys,
  widgetValue,
};
