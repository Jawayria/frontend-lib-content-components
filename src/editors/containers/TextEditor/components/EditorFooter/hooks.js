import { useSelector } from 'react-redux';

import { RequestKeys } from '../../../../data/constants/requests';
import { selectors, thunkActions } from '../../../../data/redux';
import * as textEditorHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = textEditorHooks;

export const handleSaveClicked = ({ dispatch, editorRef }) => () => saveBlock({
  editorRef,
  saveBlockContent: (...args) => dispatch(thunkActions.app.saveBlock(...args)),
  returnUrl: useSelector(selectors.app.returnUrl),
});

export const handleCancelClicked = () => navigateCallback(useSelector(selectors.app.returnUrl));

export const isInitialized = () => useSelector(selectors.app.isInitialized);

export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));

export const studioEndpointUrl = () => useSelector(selectors.app.studioEndpointUrl);
