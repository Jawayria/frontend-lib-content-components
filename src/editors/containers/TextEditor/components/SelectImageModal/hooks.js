import React from 'react';
import { useDispatch } from 'react-redux';

import { thunkActions } from '../../../../data/redux';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  setSelection,
  searchSortProps,
}) => {
  const dispatch = useDispatch();
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    dispatch(thunkActions.app.fetchImages({ setImages }));
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ setSelection }) => {
  const dispatch = useDispatch();
  const ref = React.useRef();
  const click = () => ref.current.click();
  const addFile = (e) => {
    dispatch(thunkActions.app.uploadImage({
      file: e.target.files[0],
      setSelection,
    }));
  };

  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ setSelection });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
