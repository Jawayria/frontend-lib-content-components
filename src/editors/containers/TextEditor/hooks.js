import { useState } from 'react';
import * as module from './hooks';

export const addImageUploadButton = (openModal, setImage) => (editor) => {
  editor.ui.registry.addButton('imageuploadbutton', {
    icon: 'image',
    onAction: openModal,
  });
  editor.ui.registry.addButton('editimagesettings', {
    icon: 'image',
    onAction() {
      const imgHTML = editor.selection.getNode();
      setImage({
        externalUrl: imgHTML.src,
        altText: imgHTML.alt,
        width: imgHTML.width,
        height: imgHTML.height,
      });
      openModal();
    },
  });
};

export const initializeEditorRef = (setRef, initializeEditor) => (editor) => {
  setRef(editor);
  initializeEditor();
};

// for toast onClose to avoid console warnings
export const nullMethod = () => {};

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => module.initializeEditorRef(setEditorRef, initializeEditor)(editor),
  initialValue: blockValue ? blockValue.data.data : '',
  init: {
    setup: module.addImageUploadButton(openModal, setSelection),
    plugins: 'link codesample emoticons table charmap code autoresize image imagetools',
    menubar: false,
    toolbar: 'undo redo | formatselect | '
      + 'bold italic backcolor | alignleft aligncenter '
      + 'alignright alignjustify | bullist numlist outdent indent |'
      + 'imageuploadbutton | link | emoticons | table | codesample | charmap |'
      + 'removeformat | hr |code',
    imagetools_toolbar: 'rotateleft rotateright | flipv fliph | editimagesettings',
    imagetools_cors_hosts: ['courses.edx.org'],
    height: '100%',
    content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:14px }',
    min_height: 1000,
    branding: false,
  },
});

export const modalToggle = () => {
  const [isOpen, setIsOpen] = useState(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};
