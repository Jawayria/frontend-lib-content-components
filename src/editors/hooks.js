import { useEffect } from 'react';

import * as module from './hooks';

export const initializeApp = ({ initialize, data }) => useEffect(
  () => initialize(data),
  [],
);

export const navigateTo = (destination) => {
  window.location.assign(destination);
};

export const navigateCallback = (destination) => () => module.navigateTo(destination);
