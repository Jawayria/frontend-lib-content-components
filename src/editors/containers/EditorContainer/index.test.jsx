import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';

import { EditorContainer } from '.';
import * as hooks from './hooks';

const props = {
  getContent: jest.fn().mockName('props.getContent'),
  onClose: jest.fn().mockName('props.onClose'),
};

jest.mock('./hooks', () => ({
  isInitialized: jest.fn().mockReturnValue(true),
  handleCancelClicked: (args) => ({ handleCancelClicked: args }),
  handleSaveClicked: (args) => ({ handleSaveClicked: args }),
  saveFailed: jest.fn().mockName('hooks.saveFailed'),
}));

let el;

describe('EditorContainer component', () => {
  describe('render', () => {
    const testContent = (<h1>My test content</h1>);
    test('snapshot: not initialized. disable save and pass to header', () => {
      hooks.isInitialized.mockReturnValueOnce(false);
      expect(
        shallow(<EditorContainer {...props}>{testContent}</EditorContainer>),
      ).toMatchSnapshot();
    });
    test('snapshot: initialized. enable save and pass to header', () => {
      expect(
        shallow(<EditorContainer {...props}>{testContent}</EditorContainer>),
      ).toMatchSnapshot();
    });
    describe('behavior inspection', () => {
      beforeEach(() => {
        el = shallow(<EditorContainer {...props}>{testContent}</EditorContainer>);
      });
      test('close behavior is linked to modal onClose and footer onCancel', () => {
        const expected = hooks.handleCancelClicked({ onClose: props.onClose });
        expect(el.children().at(2).props().onCancel).toEqual(expected);
      });
      test('save behavior is linked to footer onSave', () => {
        const expected = hooks.handleSaveClicked({
          getContent: props.getContent,
          dispatch: useDispatch(),
        });
        expect(el.children().at(2).props().onSave).toEqual(expected);
      });
    });
  });
});
