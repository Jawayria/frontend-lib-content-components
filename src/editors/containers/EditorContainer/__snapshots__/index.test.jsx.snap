// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContainer component render snapshot: initialized. enable save and pass to header 1`] = `
<FullscreenModal
  footerNode={
    <InjectIntl(ShimmedIntlComponent)
      disableSave={false}
      onCancel={
        Object {
          "handleCancelClicked": Object {
            "onClose": [MockFunction props.onClose],
          },
        }
      }
      onSave={
        Object {
          "handleSaveClicked": Object {
            "dispatch": [MockFunction react-redux.dispatch],
            "getContent": [MockFunction props.getContent],
          },
        }
      }
    />
  }
  isOpen={true}
  onClose={
    Object {
      "handleCancelClicked": Object {
        "onClose": [MockFunction props.onClose],
      },
    }
  }
  title={
    <InjectIntl(ShimmedIntlComponent)
      isInitialized={true}
    />
  }
  variant="primary"
>
  <h1>
    My test content
  </h1>
</FullscreenModal>
`;

exports[`EditorContainer component render snapshot: not initialized. disable save and pass to header 1`] = `
<FullscreenModal
  footerNode={
    <InjectIntl(ShimmedIntlComponent)
      disableSave={true}
      onCancel={
        Object {
          "handleCancelClicked": Object {
            "onClose": [MockFunction props.onClose],
          },
        }
      }
      onSave={
        Object {
          "handleSaveClicked": Object {
            "dispatch": [MockFunction react-redux.dispatch],
            "getContent": [MockFunction props.getContent],
          },
        }
      }
    />
  }
  isOpen={true}
  onClose={
    Object {
      "handleCancelClicked": Object {
        "onClose": [MockFunction props.onClose],
      },
    }
  }
  title={
    <InjectIntl(ShimmedIntlComponent)
      isInitialized={false}
    />
  }
  variant="primary"
>
  <h1>
    My test content
  </h1>
</FullscreenModal>
`;
